helix-core/src/main/java/org/apache/helix/cloud/topology/VirtualTopologyGroupScheme.java

```diff
@@ -0,0 +1,73 @@
+package org.apache.helix.cloud.topology;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.cloud.constants.VirtualTopologyGroupConstants;
+import org.apache.helix.util.HelixUtil;
+
+public enum VirtualTopologyGroupScheme implements VirtualGroupAssignmentAlgorithm {
+
+  /**
+   * A strategy that densely assign virtual groups with input instance list, it doesn't move to the next one until
+   * the current one is filled.
+   * Given that instances.size = instancesPerGroup * numGroups + residuals,
+   * we break [residuals] into the first few groups, as a result each virtual group will have
+   * either [instancesPerGroup] or [instancesPerGroup + 1] instances.
+   */
+  FIFO {
```

Comment:
I'm a bit confused about the code structure here. You have an enum with one value only, do you plan to add more in the future? Then will the function `computeVirtualGroupId` work for all future scheme? I think you should have one class for one type of assigning scheme, which implements the interface.

Reply:
Thanks for the comment.
My original idea is to use Enum to achieve singleton pattern while implementing an interface. Future impl can be added as enum value.
However, seems like it's causing more confusion than benefits, let me create standalone impl class instead.